OpenRA.Game/WorldUtils.cs

```diff
@@ -41,6 +41,116 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			// Then we iterate over this list, and find all actors for which their health radius is at least within rayWidth of the ray.
+			// For actors without a health radius, we simply check their center point.
+			// The square in which we select all actors must be large enough to encompass the entire ray's width.
+			var xOver = 0;
+			var yOver = 0;
+			if (target.X < source.X)
+				xOver = -1024;
+			else
+				xOver = 1024;
+
+			if (target.Y < source.Y)
+				yOver = -1024;
+			else
+				yOver = 1024;
+
+			WVec dir = new WVec(xOver, yOver, 0);
+			var overselect = dir * (1 + ((int)rayWidth.Length / 1024));
+			var finalTarget = target + overselect;
+			var finalSource = source - overselect;
+
+			var actorsInSquare = world.ActorMap.ActorsInBox(finalTarget, finalSource);
+			var intersectedActors = new List<Actor>();
+
+			foreach (var currActor in actorsInSquare)
+			{
+				long actorWidth;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+				{
+					actorWidth = healthInfo.Radius.Length;
+				}
+				else
+					actorWidth = 0;
+
+				var distance = MinimumDistancePointAndVector(target, source, currActor.CenterPosition);
+				var maxReach = (long)(actorWidth + rayWidth.Length);
+				maxReach *= maxReach;
+				if (distance <= maxReach)
+				{
+					intersectedActors.Add(currActor);
+				}
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Returns the (squared) minimum distance between a line and a point.
+		/// From http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+		/// To avoid floats, it returns the square of the distance.
+		/// </summary>
+		/// <param name="target">The target point (head) of the line</param>
+		/// <param name="source">The source point (tail) of the line</param>
+		/// <param name="point">The point that the minimum distance should be found to</param>
+		/// <returns>The square of the minimum distance between the line and point.</returns>
+		private static long MinimumDistancePointAndVector(WPos target, WPos source, WPos point)
+		{
+			// Return minimum distance between line segment vw and point p
+			var squaredLength = SquaredDistance(target, source);
+
+			// target = source case, so just find distance from the point to the target.
+			if (squaredLength == 0.0) return SquaredDistance(target, point);
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			double t = DotProduct(target, source, point);
+
+			// Beyond the 'target' end of the segment
+			if (t < 0.0)
+				return SquaredDistance(point, target);
+
+			// Beyond the 'source end of the segment
+			if (t > 1.0)
+				return SquaredDistance(point, source);
+
+			// Projection falls on the segment
+			var newX = (int)((double)target.X + t * (double)(source.X - target.X));
```

Comment:
You can't use floating point math anywhere in the world simulation code.  This method will need to be rewritten using integer math (I suggest rescaling everything so that 1.0f = 1024 to match what we do in other places).